gateway/src/error.rs

```diff
@@ -0,0 +1,42 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Error handling facilities for the management gateway.
+
+use crate::http_entrypoints::SpIdentifier;
+use dropshot::HttpError;
+use serde::{Deserialize, Serialize};
+
+/// An error that can be generated within the gateway.
+#[derive(Debug, Deserialize, thiserror::Error, PartialEq, Serialize)]
+pub(crate) enum Error {
+    /// A requested SP does not exist.
+    ///
+    /// This is not the same as the requested SP target being offline; this
+    /// error indicates a fatal, invalid request (e.g., asing for the SP on the
+    /// 17th switch when there are only two switches).
```

Comment:
This is an interesting concept. Let's saw a customer had a half-populated rack. It sounds like we'll treat a request to, say, sled 20 different than to sled 200. What if the customer is in the process of adding 16 sleds? I don't have a proposal here. It does seem like the gateway service on startup will decide on the topology of its rack so can tell the difference between "sled 20 might some day connect to something" and "sled 200 will never connect to anything". Would a client (nexus) handle these differently?

Perhaps this also suggests that the gateway should have a "give me your topology" endpoint? What do you think?

Reply:
> It does seem like the gateway service on startup will decide on the topology of its rack so can tell the difference between "sled 20 might some day connect to something" and "sled 200 will never connect to anything". 

This is exactly what I was thinking when I wrote this comment. Things get very hazy for me at this point, but I don't think nexus should ever be sending requests to MGS for slots that will never exist - how it knows that is a good question. If MGS is the thing that determines the topology of the rack (populated or not) then a "give my your topo" endpoint makes sense to me, but maybe that's determined by RSS and provided to MGS during startup?